snuba/migrations/snuba_migrations/sessions/0003_sessions_matview.py

```diff
@@ -0,0 +1,139 @@
+
+    -- use the old `count` columns for "legacy" rows with `quantity == 0`:
+    countIfState(session_id, seq == 0 AND quantity == 0) as sessions,
+    countIfState(session_id, status == 2 AND quantity == 0) as sessions_crashed,
+    countIfState(session_id, status == 3 AND quantity == 0) as sessions_abnormal,
+
+    -- for new sessions that have a `quantity`, we use `sum`:
+    sumIfState(quantity, seq == 0 AND quantity > 0) as sessions_preaggr,
+    sumIfState(quantity, status == 2 AND quantity > 0) as sessions_crashed_preaggr,
+    sumIfState(quantity, status == 3 AND quantity > 0) as sessions_abnormal_preaggr,
+
+    -- individual session updates keep using the uniq session_id as before:
+    uniqIfState(session_id, errors > 0 AND session_id != '{NIL_UUID}') as sessions_errored,
+
+    -- pre-aggregated counts use sum. by definition, `crashed` and `abnormal` are errored:
+    sumIfState(quantity, status IN (2, 3, 4) AND session_id == '{NIL_UUID}') as sessions_errored_preaggr,
+
+    -- users counts will additionally be constrained for the distinct-id:
+    uniqIfState(distinct_id, distinct_id != '{NIL_UUID}') as users,
+    uniqIfState(distinct_id, status == 2 AND distinct_id != '{NIL_UUID}') as users_crashed,
+    uniqIfState(distinct_id, status == 3 AND distinct_id != '{NIL_UUID}') as users_abnormal,
+    uniqIfState(distinct_id, errors > 0 AND distinct_id != '{NIL_UUID}') as users_errored
+
+FROM
+     sessions_raw_local
+GROUP BY
+     org_id, project_id, started, release, environment, user_agent, os, runtime
+"""
+
+
+class Migration(migration.MultiStepMigration):
+    """
+    This migration re-creates the materialized view that aggregates sessions.
+    It is using either the old, or the new `X_preaggr` columns based on the
+    `quantity`, which defaults to `0`, but is guaranteed to be `>= 1` once the
+    processor is updated.
+    """
+
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_mv_local",
+            ),
+            operations.CreateMaterializedView(
+                storage_set=StorageSetKey.SESSIONS,
+                view_name="sessions_hourly_mv_local",
+                destination_table_name="sessions_hourly_local",
+                columns=aggregate_columns,
+                query=query,
+            ),
+        ]
+
+    def backwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.DropTable(
+                storage_set=StorageSetKey.SESSIONS,
+                table_name="sessions_hourly_mv_local",
+            ),
+            create_matview_v1,
+        ]
+
+    def forwards_dist(self) -> Sequence[operations.Operation]:
+        return []
+
+    def backwards_dist(self) -> Sequence[operations.Operation]:
+        return []
```

Comment:
missing

Reply:
Are these needed for the matview itself? The initial `0001` migration does not have these either.

Reply:
ok, ignore then. I thought we had the DDL for the matview in the distributed migration.